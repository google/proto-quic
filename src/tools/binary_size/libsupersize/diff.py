# Copyright 2017 The Chromium Authors. All rights reserved.
# Use of this source code is governed by a BSD-style license that can be
# found in the LICENSE file.
"""Logic for diffing two SizeInfo objects."""

import collections
import re

import models


def _SymbolKey(symbol):
  """Returns a tuple that can be used to see if two Symbol are the same.

  Keys are not guaranteed to be unique within a SymbolGroup. When multiple
  symbols have the same key, they will be matched up in order of appearance.
  We do this because the numbering of these generated symbols is not stable.

  Examples of symbols with shared keys:
    "** merge strings"
    "** symbol gap 3", "** symbol gap 5"
    "foo() [clone ##]"
    "CSWTCH.61", "CSWTCH.62"
    "._468", "._467"
    ".L__unnamed_1193", ".L__unnamed_712"
  """
  name = symbol.full_name
  clone_idx = name.find(' [clone ')
  if clone_idx != -1:
    name = name[:clone_idx]
  if name.startswith('*'):
    # "symbol gap 3 (bar)" -> "symbol gaps"
    name = re.sub(r'\s+\d+( \(.*\))?$', 's', name)

  # Use section rather than section_name since clang & gcc use
  # .data.rel.ro vs .data.rel.ro.local.
  if '.' not in name:
    return (symbol.section, name)

  # Compiler or Linker generated symbol.
  name = re.sub(r'[.0-9]', '', name)  # Strip out all numbers and dots.
  return (symbol.section, name, symbol.object_path)


def _DiffSymbolGroups(before, after):
  before_symbols_by_key = collections.defaultdict(list)
  for s in before:
    before_symbols_by_key[_SymbolKey(s)].append(s)

  delta_symbols = []
  # For changed symbols, padding is zeroed out. In order to not lose the
  # information entirely, store it in aggregate.
  padding_by_section_name = collections.defaultdict(int)

  # Create a DeltaSymbol for each after symbol.
  for after_sym in after:
    matching_syms = before_symbols_by_key.get(_SymbolKey(after_sym))
    before_sym = None
    if matching_syms:
      before_sym = matching_syms.pop(0)
      # Padding tracked in aggregate, except for padding-only symbols.
      if before_sym.size_without_padding:
        padding_by_section_name[before_sym.section_name] += (
            after_sym.padding_pss - before_sym.padding_pss)
    delta_symbols.append(models.DeltaSymbol(before_sym, after_sym))

  # Create a DeltaSymbol for each unmatched before symbol.
  for remaining_syms in before_symbols_by_key.itervalues():
    for before_sym in remaining_syms:
      delta_symbols.append(models.DeltaSymbol(before_sym, None))

  # Create a DeltaSymbol to represent the zero'd out padding of matched symbols.
  for section_name, padding in padding_by_section_name.iteritems():
    if padding != 0:
      after_sym = models.Symbol(section_name, padding,
                                name="** aggregate padding of diff'ed symbols")
      after_sym.padding = padding
      delta_symbols.append(models.DeltaSymbol(None, after_sym))

  return models.DeltaSymbolGroup(delta_symbols)


def Diff(before, after):
  """Diffs two SizeInfo objects. Returns a DeltaSizeInfo."""
  assert isinstance(before, models.SizeInfo)
  assert isinstance(after, models.SizeInfo)
  section_sizes = {k: after.section_sizes.get(k, 0) - v
                   for k, v in before.section_sizes.iteritems()}
  for k, v in after.section_sizes.iteritems():
    if k not in section_sizes:
      section_sizes[k] = v

  symbol_diff = _DiffSymbolGroups(before.raw_symbols, after.raw_symbols)
  return models.DeltaSizeInfo(section_sizes, symbol_diff, before.metadata,
                              after.metadata)
