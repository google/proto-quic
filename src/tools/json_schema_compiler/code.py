# Copyright (c) 2012 The Chromium Authors. All rights reserved.
# Use of this source code is governed by a BSD-style license that can be
# found in the LICENSE file.

class Code(object):
  """A convenience object for constructing code.

  Logically each object should be a block of code. All methods except |Render|
  and |IsEmpty| return self.
  """
  def __init__(self, indent_size=2, comment_length=80):
    self._code = []
    self._indent_size = indent_size
    self._comment_length = comment_length
    self._line_prefixes = []

  def Append(self, line='',
             substitute=True,
             indent_level=None,
             new_line=True,
             strip_right=True):
    """Appends a line of code at the current indent level or just a newline if
    line is not specified.

    substitute: indicated whether this line should be affected by
    code.Substitute().
    new_line: whether this should be added as a new line, or should be appended
        to the last line of the code.
    strip_right: whether or not trailing whitespace should be stripped.
    """

    if line:
      prefix = indent_level * ' ' if indent_level else ''.join(
          self._line_prefixes)
    else:
      prefix = ''

    if strip_right:
      line = line.rstrip()

    if not new_line and self._code:
      self._code[-1].value += line
    else:
      self._code.append(Line(prefix + line, substitute=substitute))
    return self

  def IsEmpty(self):
    """Returns True if the Code object is empty.
    """
    return not bool(self._code)

  def Concat(self, obj, new_line=True):
    """Concatenate another Code object onto this one. Trailing whitespace is
    stripped.

    Appends the code at the current indent level. Will fail if there are any
    un-interpolated format specifiers eg %s, %(something)s which helps
    isolate any strings that haven't been substituted.
    """
    if not isinstance(obj, Code):
      raise TypeError(type(obj))
    assert self is not obj
    if not obj._code:
      return self

    for line in obj._code:
      try:
        # line % () will fail if any substitution tokens are left in line
        if line.substitute:
          line.value %= ()
      except TypeError:
        raise TypeError('Unsubstituted value when concatting\n' + line.value)
      except ValueError:
        raise ValueError('Stray % character when concatting\n' + line.value)
    first_line = obj._code.pop(0)
    self.Append(first_line.value, first_line.substitute, new_line=new_line)
    for line in obj._code:
      self.Append(line.value, line.substitute)

    return self

  def Cblock(self, code):
    """Concatenates another Code object |code| onto this one followed by a
    blank line, if |code| is non-empty."""
    if not code.IsEmpty():
      self.Concat(code).Append()
    return self

  def Sblock(self, line=None, line_prefix=None, new_line=True):
    """Starts a code block.

    Appends a line of code and then increases the indent level. If |line_prefix|
    is present, it will be treated as the extra prefix for the code block.
    Otherwise, the prefix will be the default indent level.
    """
    if line is not None:
      self.Append(line, new_line=new_line)
    self._line_prefixes.append(line_prefix or ' ' * self._indent_size)
    return self

  def Eblock(self, line=None):
    """Ends a code block by decreasing and then appending a line (or a blank
    line if not given).
    """
    # TODO(calamity): Decide if type checking is necessary
    #if not isinstance(line, basestring):
    #  raise TypeError
    self._line_prefixes.pop()
    if line is not None:
      self.Append(line)
    return self

  def Comment(self, comment, comment_prefix='// ',
              wrap_indent=0, new_line=True):
    """Adds the given string as a comment.

    Will split the comment if it's too long. Use mainly for variable length
    comments. Otherwise just use code.Append('// ...') for comments.

    Unaffected by code.Substitute().
    """
    # Helper function to trim a comment to the maximum length, and return one
    # line and the remainder of the comment.
    def trim_comment(comment, max_len):
      if len(comment) <= max_len:
        return comment, ''
      last_space = comment.rfind(' ', 0, max_len + 1)
      if last_space != -1:
        line = comment[0:last_space]
        comment = comment[last_space + 1:]
      else:
        line = comment[0:max_len]
        comment = comment[max_len:]
      return line, comment.lstrip()

    # First line has the full maximum length.
    if not new_line and self._code:
      max_len = self._comment_length - len(self._code[-1].value)
    else:
      max_len = (self._comment_length - len(''.join(self._line_prefixes)) -
                 len(comment_prefix))
    line, comment = trim_comment(comment, max_len)
    self.Append(comment_prefix + line, substitute=False, new_line=new_line)

    # Any subsequent lines be subject to the wrap indent.
    max_len = (self._comment_length - len(''.join(self._line_prefixes)) -
               len(comment_prefix) - wrap_indent)
    while len(comment):
      line, comment = trim_comment(comment, max_len)
      self.Append(comment_prefix + (' ' * wrap_indent) + line, substitute=False)

    return self

  def Substitute(self, d):
    """Goes through each line and interpolates using the given dict.

    Raises type error if passed something that isn't a dict

    Use for long pieces of code using interpolation with the same variables
    repeatedly. This will reduce code and allow for named placeholders which
    are more clear.
    """
    if not isinstance(d, dict):
      raise TypeError('Passed argument is not a dictionary: ' + d)
    for i, line in enumerate(self._code):
      if self._code[i].substitute:
        # Only need to check %s because arg is a dict and python will allow
        # '%s %(named)s' but just about nothing else
        if '%s' in self._code[i].value or '%r' in self._code[i].value:
          raise TypeError('"%s" or "%r" found in substitution. '
                          'Named arguments only. Use "%" to escape')
        self._code[i].value = line.value % d
        self._code[i].substitute = False
    return self

  def TrimTrailingNewlines(self):
    """Removes any trailing empty Line objects.
    """
    while self._code:
      if self._code[-1].value != '':
        return
      self._code = self._code[:-1]

  def Render(self):
    """Renders Code as a string.
    """
    return '\n'.join([l.value for l in self._code])


class Line(object):
  """A line of code.
  """
  def __init__(self, value, substitute=True):
    self.value = value
    self.substitute = substitute
