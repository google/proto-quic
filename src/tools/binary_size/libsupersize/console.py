# Copyright 2017 The Chromium Authors. All rights reserved.
# Use of this source code is governed by a BSD-style license that can be
# found in the LICENSE file.

"""An interactive console for looking analyzing .size files."""

import argparse
import atexit
import code
import contextlib
import itertools
import logging
import os
import readline
import subprocess
import sys

import archive
import canned_queries
import describe
import diff
import file_format
import match_util
import models
import paths


# Number of lines before using less for Print().
_THRESHOLD_FOR_PAGER = 50


@contextlib.contextmanager
def _LessPipe():
  """Output to `less`. Yields a file object to write to."""
  try:
    proc = subprocess.Popen(['less'], stdin=subprocess.PIPE, stdout=sys.stdout)
    yield proc.stdin
    proc.stdin.close()
    proc.wait()
  except IOError:
    pass  # Happens when less is quit before all data is written.
  except KeyboardInterrupt:
    pass  # Assume used to break out of less.


def _WriteToStream(lines, use_pager=None, to_file=None):
  if to_file:
    use_pager = False
  if use_pager is None and sys.stdout.isatty():
    # Does not take into account line-wrapping... Oh well.
    first_lines = list(itertools.islice(lines, _THRESHOLD_FOR_PAGER))
    if len(first_lines) == _THRESHOLD_FOR_PAGER:
      use_pager = True
    lines = itertools.chain(first_lines, lines)

  if use_pager:
    with _LessPipe() as stdin:
      describe.WriteLines(lines, stdin.write)
  elif to_file:
    with open(to_file, 'w') as file_obj:
      describe.WriteLines(lines, file_obj.write)
  else:
    describe.WriteLines(lines, sys.stdout.write)


class _Session(object):
  _readline_initialized = False

  def __init__(self, size_infos, lazy_paths):
    self._printed_variables = []
    self._variables = {
        'Print': self._PrintFunc,
        'Diff': self._DiffFunc,
        'Disassemble': self._DisassembleFunc,
        'ExpandRegex': match_util.ExpandRegexIdentifierPlaceholder,
        'ShowExamples': self._ShowExamplesFunc,
        'canned_queries': canned_queries.CannedQueries(size_infos),
        'printed': self._printed_variables,
    }
    self._lazy_paths = lazy_paths
    self._size_infos = size_infos
    self._disassemble_prefix_len = None

    if len(size_infos) == 1:
      self._variables['size_info'] = size_infos[0]
    else:
      for i, size_info in enumerate(size_infos):
        self._variables['size_info%d' % (i + 1)] = size_info

  def _DiffFunc(self, before=None, after=None, sort=True):
    """Diffs two SizeInfo objects. Returns a DeltaSizeInfo.

    Args:
      before: Defaults to first size_infos[0].
      after: Defaults to second size_infos[1].
      sort: When True (default), calls SymbolGroup.Sorted() after diffing.
    """
    before = before if before is not None else self._size_infos[0]
    after = after if after is not None else self._size_infos[1]
    ret = diff.Diff(before, after)
    if sort:
      ret.symbols = ret.symbols.Sorted()
    return ret

  def _PrintFunc(self, obj=None, verbose=False, recursive=False, use_pager=None,
                 to_file=None):
    """Prints out the given Symbol / SymbolGroup / SizeInfo.

    For convenience, |obj| will be appended to the global "printed" list.

    Args:
      obj: The object to be printed. Defaults to size_infos[-1]. Also accepts an
          index into the |printed| array for showing previous results.
      verbose: Show more detailed output.
      recursive: Print children of nested SymbolGroups.
      use_pager: Pipe output through `less`. Ignored when |obj| is a Symbol.
          default is to automatically pipe when output is long.
      to_file: Rather than print to stdio, write to the given file.
    """
    if isinstance(obj, int):
      obj = self._printed_variables[obj]
    elif not self._printed_variables or self._printed_variables[-1] != obj:
      if not isinstance(obj, models.SymbolGroup) or len(obj) > 0:
        self._printed_variables.append(obj)
    obj = obj if obj is not None else self._size_infos[-1]
    lines = describe.GenerateLines(obj, verbose=verbose, recursive=recursive)
    _WriteToStream(lines, use_pager=use_pager, to_file=to_file)

  def _ElfPathAndToolPrefixForSymbol(self, size_info, elf_path):
    tool_prefix = self._lazy_paths.tool_prefix
    orig_tool_prefix = size_info.metadata.get(models.METADATA_TOOL_PREFIX)
    if orig_tool_prefix:
      orig_tool_prefix = paths.FromSrcRootRelative(orig_tool_prefix)
      if os.path.exists(orig_tool_prefix + 'objdump'):
        tool_prefix = orig_tool_prefix

    # TODO(agrieve): Would be even better to use objdump --info to check that
    #     the toolchain is for the correct architecture.
    assert tool_prefix is not None, (
        'Could not determine --tool-prefix. Possible fixes include setting '
        '--tool-prefix, or setting --output-directory')

    def build_id_matches(elf_path):
      found_build_id = archive.BuildIdFromElf(elf_path, tool_prefix)
      expected_build_id = size_info.metadata.get(models.METADATA_ELF_BUILD_ID)
      return found_build_id == expected_build_id

    filename = size_info.metadata.get(models.METADATA_ELF_FILENAME)
    paths_to_try = []
    if elf_path:
      paths_to_try.append(elf_path)
    else:
      auto_lazy_paths = [
          paths.LazyPaths(any_path_within_output_directory=s.size_path)
          for s in self._size_infos]
      for lazy_paths in auto_lazy_paths + [self._lazy_paths]:
        output_dir = lazy_paths.output_directory
        if output_dir:
          # Local build: File is located in output directory.
          paths_to_try.append(
              os.path.normpath(os.path.join(output_dir, filename)))
        # Downloaded build: File is located beside .size file.
        paths_to_try.append(os.path.normpath(os.path.join(
            os.path.dirname(size_info.size_path), os.path.basename(filename))))

    paths_to_try = [p for p in paths_to_try if os.path.exists(p)]

    for i, elf_path in enumerate(paths_to_try):
      if build_id_matches(elf_path):
        return elf_path, tool_prefix

      # Show an error only once all paths are tried.
      if i + 1 == len(paths_to_try):
        assert False, 'Build ID does not match for %s' % elf_path

    assert False, (
        'Could not locate ELF file. If binary was built locally, ensure '
        '--output-directory is set. If output directory is unavailable, '
        'ensure {} is located beside {}, or pass its path explicitly using '
        'elf_path=').format(os.path.basename(filename), size_info.size_path)

  def _DetectDisassemblePrefixLen(self, args):
    # Look for a line that looks like:
    # /usr/{snip}/src/out/Release/../../net/quic/core/quic_time.h:100
    output = subprocess.check_output(args)
    for line in output.splitlines():
      if line and line[0] == os.path.sep and line[-1].isdigit():
        release_idx = line.find('Release')
        if release_idx != -1:
          return line.count(os.path.sep, 0, release_idx)
        dot_dot_idx = line.find('..')
        if dot_dot_idx != -1:
          return line.count(os.path.sep, 0, dot_dot_idx) - 1
        out_idx = line.find(os.path.sep + 'out')
        if out_idx != -1:
          return line.count(os.path.sep, 0, out_idx) + 2
        logging.warning('Could not guess source path from found path.')
        return None
    logging.warning('Found no source paths in objdump output.')
    return None

  def _DisassembleFunc(self, symbol, elf_path=None, use_pager=None,
                       to_file=None):
    """Shows objdump disassembly for the given symbol.

    Args:
      symbol: Must be a .text symbol and not a SymbolGroup.
      elf_path: Path to the executable containing the symbol. Required only
          when auto-detection fails.
    """
    assert not symbol.IsGroup()
    assert symbol.address and symbol.section_name == '.text'
    assert not symbol.IsDelta(), ('Cannot disasseble a Diff\'ed symbol. Try '
                                  'passing .before_symbol or .after_symbol.')
    size_info = None
    for size_info in self._size_infos:
      if symbol in size_info.raw_symbols:
        break
    else:
      assert False, 'Symbol does not belong to a size_info.'

    elf_path, tool_prefix = self._ElfPathAndToolPrefixForSymbol(
        size_info, elf_path)

    args = [tool_prefix + 'objdump', '--disassemble', '--source',
            '--line-numbers', '--demangle',
            '--start-address=0x%x' % symbol.address,
            '--stop-address=0x%x' % symbol.end_address, elf_path]
    if self._disassemble_prefix_len is None:
      prefix_len = self._DetectDisassemblePrefixLen(args)
      if prefix_len is not None:
        self._disassemble_prefix_len = prefix_len

    if self._disassemble_prefix_len is not None:
      output_directory = self._lazy_paths.output_directory
      # Only matters for non-generated paths, so be lenient here.
      if output_directory is None:
        output_directory = os.path.join(paths.SRC_ROOT, 'out', 'Release')
        if not os.path.exists(output_directory):
          os.makedirs(output_directory)

      args += [
          '--prefix-strip', str(self._disassemble_prefix_len),
          '--prefix', os.path.normpath(os.path.relpath(output_directory))
      ]

    proc = subprocess.Popen(args, stdout=subprocess.PIPE)
    lines = itertools.chain(('Showing disassembly for %r' % symbol,
                             'Command: %s' % ' '.join(args)),
                            (l.rstrip() for l in proc.stdout))
    _WriteToStream(lines, use_pager=use_pager, to_file=to_file)
    proc.kill()

  def _ShowExamplesFunc(self):
    print self._CreateBanner()
    print '\n'.join([
        '# Show pydoc for main types:',
        'import models',
        'help(models)',
        '',
        '# Show all attributes of all symbols & per-section totals:',
        'Print(size_info, verbose=True)',
        '',
        '# Show two levels of .text, grouped by first two subdirectories',
        'text_syms = size_info.symbols.WhereInSection("t")',
        'by_path = text_syms.GroupedByPath(depth=2)',
        'Print(by_path.WherePssBiggerThan(1024))',
        '',
        '# Show all generated symbols, then show only non-vtable ones',
        'Print(size_info.symbols.WhereGeneratedByToolchain())',
        'Print(printed[-1].WhereNameMatches(r"vtable").Inverted().Sorted())',
        '',
        '# Show all symbols that have "print" in their name or path, except',
        '# those within components/.',
        '# Note: Could have also used Inverted(), as above.',
        '# Note: Use "help(ExpandRegex)" for more about what {{_print_}} does.',
        'print_syms = size_info.symbols.WhereMatches(r"{{_print_}}")',
        'Print(print_syms - print_syms.WherePathMatches(r"^components/"))',
        '',
        '# Diff two .size files and save result to a file:',
        'Print(Diff(size_info1, size_info2), to_file="output.txt")',
        '',
        '# View per-component breakdowns, then drill into the last entry.',
        'c = canned_queries.CategorizeByChromeComponent()',
        'Print(c)',
        'Print(c[-1].GroupedByPath(depth=2).Sorted())',
        '',
        '# For even more inspiration, look at canned_queries.py',
        '# (and feel free to add your own!).',
    ])

  def _CreateBanner(self):
    def keys(cls, super_keys=None):
      ret = sorted(m for m in dir(cls) if m[0] != '_')
      if super_keys:
        ret = sorted(m for m in ret if m not in super_keys)
      return ret

    symbol_info_keys = keys(models.SizeInfo)
    symbol_keys = keys(models.Symbol)
    symbol_group_keys = keys(models.SymbolGroup, symbol_keys)
    delta_size_info_keys = keys(models.DeltaSizeInfo)
    delta_symbol_keys = keys(models.DeltaSymbol, symbol_keys)
    delta_symbol_group_keys = keys(models.DeltaSymbolGroup,
                                   symbol_keys + symbol_group_keys)
    canned_queries_keys = keys(canned_queries.CannedQueries)

    functions = sorted(k for k in self._variables if k[0].isupper())
    lines = [
        '*' * 80,
        'Entering interactive Python shell. Quick reference:',
        '',
        'SizeInfo: %s' % ', '.join(symbol_info_keys),
        'Symbol: %s' % ', '.join(symbol_keys),
        '',
        'SymbolGroup (extends Symbol): %s' % ', '.join(symbol_group_keys),
        '',
        'DeltaSizeInfo: %s' % ', '.join(delta_size_info_keys),
        'DeltaSymbol (extends Symbol): %s' % ', '.join(delta_symbol_keys),
        'DeltaSymbolGroup (extends SymbolGroup): %s' % ', '.join(
            delta_symbol_group_keys),
        '',
        'canned_queries: %s' % ', '.join(canned_queries_keys),
        '',
        'Functions: %s' % ', '.join('%s()' % f for f in functions),
        'Variables: ',
        '  printed: List of objects passed to Print().',
    ]
    for key, value in self._variables.iteritems():
      if key.startswith('size_info'):
        lines.append('  {}: Loaded from {}'.format(key, value.size_path))
    lines.append('*' * 80)
    return '\n'.join(lines)


  @classmethod
  def _InitReadline(cls):
    if cls._readline_initialized:
      return
    cls._readline_initialized = True
    # Without initializing readline, arrow keys don't even work!
    readline.parse_and_bind('tab: complete')
    history_file = os.path.join(os.path.expanduser('~'),
                                '.binary_size_query_history')
    if os.path.exists(history_file):
      readline.read_history_file(history_file)
    atexit.register(lambda: readline.write_history_file(history_file))

  def Eval(self, query):
    exec query in self._variables

  def GoInteractive(self):
    _Session._InitReadline()
    code.InteractiveConsole(self._variables).interact(self._CreateBanner())


def AddArguments(parser):
  parser.add_argument(
      'inputs', nargs='+',
      help='Input .size files to load. For a single file, it will be mapped to '
           'the variable "size_info". For multiple inputs, the names will be '
           'size_info1, size_info2, etc.')
  parser.add_argument('--query',
                      help='Execute the given snippet. '
                           'Example: Print(size_info)')
  parser.add_argument('--tool-prefix',
                      help='Path prefix for objdump. Required only for '
                           'Disassemble().')
  parser.add_argument('--output-directory',
                      help='Path to the root build directory. Used only for '
                           'Disassemble().')


def Run(args, parser):
  for path in args.inputs:
    if not path.endswith('.size'):
      parser.error('All inputs must end with ".size"')

  size_infos = [archive.LoadAndPostProcessSizeInfo(p) for p in args.inputs]
  lazy_paths = paths.LazyPaths(tool_prefix=args.tool_prefix,
                               output_directory=args.output_directory,
                               any_path_within_output_directory=args.inputs[0])
  session = _Session(size_infos, lazy_paths)

  if args.query:
    logging.info('Running query from command-line.')
    session.Eval(args.query)
  else:
    logging.info('Entering interactive console.')
    session.GoInteractive()
