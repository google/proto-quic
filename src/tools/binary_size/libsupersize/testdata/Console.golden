********************************************************************************
Entering interactive Python shell. Quick reference:

SizeInfo: metadata, raw_symbols, section_sizes, size_path, symbols
Symbol: FlagsString, IsBss, IsDelta, IsGeneratedByToolchain, IsGroup, address, aliases, end_address, flags, full_name, generated_source, is_anonymous, name, num_aliases, object_path, padding, padding_pss, pss, pss_without_padding, section, section_name, size, size_without_padding, source_path, template_name

SymbolGroup (extends Symbol): CountUniqueSymbols, Filter, GroupedBy, GroupedByFullName, GroupedByName, GroupedByPath, GroupedBySectionName, Inverted, IterLeafSymbols, IterUniqueSymbols, SetName, Sorted, SortedByAddress, SortedByCount, SortedByName, WhereAddressInRange, WhereFullNameMatches, WhereGeneratedByToolchain, WhereHasAnyAttribution, WhereHasPath, WhereInSection, WhereIsGroup, WhereIsTemplate, WhereMatches, WhereNameMatches, WhereObjectPathMatches, WherePathMatches, WherePssBiggerThan, WhereSizeBiggerThan, WhereSourceIsGenerated, WhereSourcePathMatches, WhereTemplateNameMatches, index, is_sorted

DeltaSizeInfo: after_metadata, before_metadata, raw_symbols, section_sizes, symbols
DeltaSymbol (extends Symbol): after_symbol, before_symbol, diff_status
DeltaSymbolGroup (extends SymbolGroup): CountsByDiffStatus, WhereDiffStatusIs, diff_status

canned_queries: CategorizeByChromeComponent, CategorizeGenerated, TemplatesByName

Functions: Diff(), Disassemble(), ExpandRegex(), Print(), ShowExamples()
Variables: 
  printed: List of objects passed to Print().
  size_info: Loaded from {redacted}
********************************************************************************
# Show pydoc for main types:
import models
help(models)

# Show all attributes of all symbols & per-section totals:
Print(size_info, verbose=True)

# Show two levels of .text, grouped by first two subdirectories
text_syms = size_info.symbols.WhereInSection("t")
by_path = text_syms.GroupedByPath(depth=2)
Print(by_path.WherePssBiggerThan(1024))

# Show all generated symbols, then show only non-vtable ones
Print(size_info.symbols.WhereGeneratedByToolchain())
Print(printed[-1].WhereNameMatches(r"vtable").Inverted().Sorted())

# Show all symbols that have "print" in their name or path, except
# those within components/.
# Note: Could have also used Inverted(), as above.
# Note: Use "help(ExpandRegex)" for more about what {{_print_}} does.
print_syms = size_info.symbols.WhereMatches(r"{{_print_}}")
Print(print_syms - print_syms.WherePathMatches(r"^components/"))

# Diff two .size files and save result to a file:
Print(Diff(size_info1, size_info2), to_file="output.txt")

# View per-component breakdowns, then drill into the last entry.
c = canned_queries.CategorizeByChromeComponent()
Print(c)
Print(c[-1].GroupedByPath(depth=2).Sorted())

# For even more inspiration, look at canned_queries.py
# (and feel free to add your own!).
Metadata:
    elf_arch=arm
    elf_build_id=WhatAnAmazingBuildId
    elf_file_name=elf
    elf_mtime={redacted}
    git_revision=abc123
    gn_args=var1=true var2="foo"
    map_file_name=../test.map
    tool_prefix=tools/binary_size/libsupersize/testdata/mock_toolchain/

Section Sizes (Total=95.6mb (100233874 bytes)):
    .ARM.exidx: 1.47mb (1536456 bytes) (1.5%)
    .bss: 1.24mb (1300456 bytes) (not included in totals)
    .data: 99.4kb (101768 bytes) (0.1%)
    .rel.dyn: 2.53mb (2655384 bytes) (2.6%)
    .rodata: 5.65mb (5927652 bytes) (5.9%)
    .strtab: 33.2mb (34841854 bytes) (34.8%)
    .symtab: 16.4mb (17166112 bytes) (17.1%)
    .text: 34.2mb (35900712 bytes) (35.8%)

Showing 45 symbols (42 unique) with total pss: 43785380 bytes
.text=34.2mb     .rodata=5.65mb     .data*=1.87mb     .bss=512kb      total=41.8mb
Number of unique paths: 9

Index | Running Total | Section@Address | PSS | Path
------------------------------------------------------------
0)         16 (0.0%)  t@0x28d900   16             base/page_allocator.cc
             _GLOBAL__sub_I_page_allocator.cc
1)         72 (0.0%)  t@0x28d910   56             base/page_allocator.cc
             _GLOBAL__sub_I_bbr_sender.cc
2)        100 (0.0%)  t@0x28d948   28             base/page_allocator.cc
             _GLOBAL__sub_I_pacing_sender.cc
3)        138 (0.0%)  t@0x28d964   38             base/page_allocator.cc
             extFromUUseMapping
4)        170 (0.0%)  t@0x28d98a   32             base/page_allocator.cc
             extFromUUseMapping
5)   35830930 (81.8%) t@Group      35830760       {no path}
             ** symbol gaps (count=3)
6)   35831378 (81.8%) t@0x28f000   448            third_party/icu/ucnv_ext.c
             ucnv_extMatchFromU
7)   35831406 (81.8%) t@0x28f1c8   28             third_party/icu/ucnv_ext.c
             _GLOBAL__sub_I_SkDeviceProfile.cpp
8)   35900530 (82.0%) t@0x28f1e0   69124          third_party/icu/ucnv_ext.c
             foo_bar
9)   35900554 (82.0%) t@0x2a0000   24 (size=48)   {no path}
             blink::ContiguousContainerBase::shrinkToFit (num_aliases=2)
10)  35900578 (82.0%) t@0x2a0000   24 (size=48)   third_party/icu/ucnv_ext.c
             BazAlias (num_aliases=2)
11)  35900582 (82.0%) t@0x2a0010   4 (size=12)    third_party/{shared}/2
             blink::ContiguousContainerBase::shrinkToFit (num_aliases=3)
12)  35900586 (82.0%) t@0x2a0010   4 (size=12)    third_party/fft_float.cc
             FooAlias (num_aliases=3)
13)  35900590 (82.0%) t@0x2a0010   4 (size=12)    third_party/fft_float.cc
             BarAlias (num_aliases=3)
14)  35900618 (82.0%) t@0x2a0020   28             third_party/container.c
             blink::ContiguousContainerBase::ContiguousContainerBase
15)  35900712 (82.0%) t@0x2a1000   94             third_party/container.c
             blink::PaintChunker::releasePaintChunks
16)  38538681 (88.0%) r@Group      2637969        {no path}
             ** merge strings (count=2)
17)  38542106 (88.0%) r@0x284d600  3425           {no path}
             ** merge constants
18)  41828221 (95.5%) r@Group      3286115        {no path}
             ** symbol gaps (count=2)
19)  41828229 (95.5%) r@0x284e364  8              base/page_allocator.cc
20)  41828273 (95.5%) r@0x284e370  44             base/page_allocator.cc
             Name
21)  41828305 (95.5%) r@0x284e398  32             third_party/container.c
             chrome::mojom::FilePatcher::Name_
22)  41828360 (95.5%) r@0x28f3450  55             third_party/paint.cc
             kAnimationFrameTimeHistogramClassPath
23)  41828364 (95.5%) r@0x28f3480  4              third_party/paint.cc
             blink::CSSValueKeywordsHash::findValueImpl::value_word_list
24)  41828420 (95.5%) d@0x2c176f0  56             third_party/icu/ucnv_ext.c
             ChromeMainDelegate [vtable]
25)  41828444 (95.5%) d@0x2c17728  24             third_party/icu/ucnv_ext.c
             chrome::mojom::FieldTrialRecorder [vtable]
26)  42618348 (97.3%) d@0x2c17740  789904         third_party/container.c
             chrome::mojom::FieldTrialRecorderProxy [vtable]
27)  42618380 (97.3%) d@0x2cd84e0  32             third_party/gvr-android-sdk/libgvr_shim_static_arm.a/libcontroller_api_impl.a_controller_api_impl.o
             .Lswitch.table.45
28)  42618388 (97.3%) d@0x2cd84f0  8              third_party/gvr-android-sdk/libgvr_shim_static_arm.a/libport_android_jni.a_jni_utils.o
             kSystemClassPrefixes
29)  42618444 (97.3%) d@0x2cd8500  56             third_party/paint.cc
             ChromeMainDelegateAndroid [vtable]
30)  42618468 (97.3%) d@0x2cd8538  24             base/page_allocator.cc
             mojo::MessageReceiver [vtable]
31)  42618480 (97.3%) d@0x2cd8550  12             base/page_allocator.cc
             kMethodsAnimationFrameTimeHistogram
32)  43683612 (99.8%) d@0x2ddc608  1065132        {no path}
             ** symbol gap 0 (end of section)
33)  43683616 (99.8%) d@0x2de7000  4              base/page_allocator.cc
             google::protobuf::internal::pLinuxKernelCmpxchg
34)  43683620 (99.8%) d@0x2de7004  4              third_party/container.c
             google::protobuf::internal::pLinuxKernelMemoryBarrier
35)  43683772 (99.8%) d@0x2de7008  152            third_party/container.c
             base::android::kBaseRegisteredMethods
36)  43683776 (99.8%) d@0x2de70a0  4              third_party/container.c
             base::android::g_renderer_histogram_code
37)  43683780 (99.8%) d@0x2de70a4  4              third_party/container.c
             base::android::g_library_version_number
38)  43785380 (100.0%) d@0x2dffd88  101600         {no path}
             ** symbol gap 0 (end of section)
39)  43785380 (100.0%) b@0x0        262144         third_party/fft_float.cc
             ff_cos_131072
40)  43785380 (100.0%) b@0x0        131072         third_party/fft_fixed.cc
             ff_cos_131072_fixed
41)  43785380 (100.0%) b@0x0        131072         third_party/fft_float.cc
             ff_cos_65536
42)  43785380 (100.0%) b@0x2dffda0  28             third_party/icu/ucnv_ext.c
             g_chrome_content_browser_client
43)  43785380 (100.0%) b@0x2dffe80  200            third_party/icu/ucnv_ext.c
             SaveHistogram::atomic_histogram_pointer
44)  43785380 (100.0%) b@0x2dffe84  4              third_party/icu/ucnv_ext.c
             g_AnimationFrameTimeHistogram_clazz
